bundles/org.openhab.binding.vigicrues/src/main/java/org/openhab/binding/vigicrues/internal/handler/VigiCruesHandler.java

```diff
@@ -54,35 +74,124 @@
 
-    public VigiCruesHandler(Thing thing, TimeZoneProvider timeZoneProvider, Gson gson) {
+    public VigiCruesHandler(Thing thing, LocationProvider locationProvider, ApiHandler apiHandler) {
         super(thing);
-        this.timeZoneProvider = timeZoneProvider;
-        this.gson = gson;
+        this.apiHandler = apiHandler;
+        this.locationProvider = locationProvider;
     }
 
     @Override
     public void initialize() {
         logger.debug("Initializing VigiCrues handler.");
 
-        VigiCruesConfiguration config = getConfigAs(VigiCruesConfiguration.class);
-        logger.debug("config station = {}", config.id);
+        StationConfiguration config = getConfigAs(StationConfiguration.class);
         logger.debug("config refresh = {} min", config.refresh);
 
         updateStatus(ThingStatus.UNKNOWN);
-        queryUrl = URL + config.id;
+
+        if (thing.getProperties().isEmpty()) {
+            Map<String, String> properties = discoverAttributes(config);
+            updateProperties(properties);
+        }
+        getReferences();
         refreshJob = scheduler.scheduleWithFixedDelay(this::updateAndPublish, 0, config.refresh, TimeUnit.MINUTES);
     }
 
+    private void getReferences() {
+        List<QuantityType<?>> heights = new ArrayList<>();
+        List<QuantityType<?>> flows = new ArrayList<>();
+        thing.getProperties().keySet().stream().filter(key -> key.startsWith(FLOOD)).forEach(key -> {
+            String value = thing.getProperties().get(key);
+            if (key.contains(FLOW)) {
+                flows.add(new QuantityType<>(value));
+            } else {
+                heights.add(new QuantityType<>(value));
+            }
+        });
+        referenceHeights = heights.stream().distinct().sorted().collect(Collectors.toList());
+        referenceFlows = flows.stream().distinct().sorted().collect(Collectors.toList());
+        portion = thing.getProperties().get(TRONCON);
+    }
+
+    private Map<String, String> discoverAttributes(StationConfiguration config) {
+        Map<String, String> properties = new HashMap<>();
+
+        ThingBuilder thingBuilder = editThing();
+        List<Channel> channels = new ArrayList<>(getThing().getChannels());
+
+        try {
+            HubEauResponse stationDetails = apiHandler.discoverStations(config.id);
+            stationDetails.stations.stream().findFirst().ifPresent(station -> {
+                PointType stationLocation = new PointType(
+                        String.format(Locale.US, "%f,%f", station.latitudeStation, station.longitudeStation));
+                properties.put(LOCATION, stationLocation.toString());
+                PointType serverLocation = locationProvider.getLocation();
+                if (serverLocation != null) {
+                    DecimalType distance = serverLocation.distanceFrom(stationLocation);
+                    properties.put(DISTANCE, new QuantityType<>(distance.intValue(), SIUnits.METRE).toString());
```

Comment:
IIRC QuantityType accepts a plain Number argument and DecimalType extends Number.
```suggestion
                    properties.put(DISTANCE, new QuantityType<>(distance, SIUnits.METRE).toString());
```